import "tensorflow/core/framework/tensor.proto";
import "tensorflow/core/protobuf/config.proto";

package nexus.turing;

option cc_generic_services = true;
option cc_enable_arenas = true;

enum CompressType {
    CT_NO_COMPRESS = 0;
    CT_Z_SPEED_COMPRESS = 1;
    CT_Z_DEFAULT_COMPRESS = 2;
    CT_SNAPPY = 3;
}

message NamedTensorProto {
  optional string name = 1;
  optional tensorflow.TensorProto tensor = 2;
}

message NamedRunMetadata {
    optional string name = 1;
    optional tensorflow.RunMetadata runMetadata = 2;
}

message GraphInfo {
    repeated NamedTensorProto inputs = 1;
    repeated string targets = 2;
    repeated string fetches = 3;
}

message GraphRequest {
    optional string bizName = 1;
    optional CompressType compressType = 2 [default = CT_NO_COMPRESS];
    optional int64 timeout = 3;
    optional string src = 4;
    optional GraphInfo graphInfo = 5;
    optional tensorflow.RunOptions runOptions = 6;
}

message GraphResponse {
    repeated NamedTensorProto outputs = 1;
    optional int32 weightInfo = 3;
    optional string debugInfo = 4;
    repeated NamedRunMetadata runMetas = 12;
}

// message BizGraphRequest {
//     required string bizName = 1;
// }

// message BizGraphResponse {
//     optional string searchGraphDef = 1;
//     optional string proxyGraphDef = 2;
//     optional string graphDef = 4;
// }

service GraphService {
    rpc runGraph(GraphRequest) returns(GraphResponse);
    // rpc getBizGraph(BizGraphRequest) returns(BizGraphResponse);
}